### DISCUSSION TITLE
Ability to bulk edit field values on multiple entries/assets at once

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8132] [Ideas]
> Created by: Emyr Thomas (emyr.thomas@gmail.com) on 2016/01/10 23:16:08 +0000
> Votes at time of UserVoice import: 15

-----------------------------------

It would be a huge time-saver if Craft has the ability to edit multiple items at once. Currently, you can bulk delete elements on the Entries index listing in the Control Panel by selecting multiple entries and selecting Delete from the Actions dropdown, but it would be very useful to change the values of any fields the selected entries have in common.

As an example, I have added an Artist (Element) field to Assets in my project. When I bulk upload a ton of images, it would be great if I could select multiple assets and edit the Artist field for each in one go.

WordPress has an implementation of this kind of functionality, see https://en.support.wordpress.com/posts/edit-posts-screen/#bulk-edit

**Comment by angrybrad** [comment]
> Posted by Christopher Robert Kennedy (chris@onebrightlight.com) on 2015/03/20 22:52:33 +0000
-----------------------------------

Oh my god, yes. I can&#39;t tell you how much time it takes up when trying to change large groups of category fields. :-/

**Comment by angrybrad** [comment]
> Posted by Timur (tim@timdov.com) on 2016/03/17 15:01:10 +0000
-----------------------------------

Great idea. I&#39;ve made similar request:
https://github.com/craftcms/cms/issues/1069

**Comment by blacksaltstudio** [comment]
This would be amazing!!!!!! +1

**Comment by pierreluckffein** [comment]
This plugin does it for Craft 3: https://github.com/kffein/craft-bulk-edit

**Comment by j-greig** [comment]
Would love this feature - for Commerce products too!

**Comment by RitterKnightCreative** [comment]
Had a client asking for this today.  Will try the plugin but this definitely needs to be in core!

**Comment by brandonkelly** [comment]
Added to the 3.3 milestone.

**Comment by blacksaltstudio** [comment]
Will this be for Assets as well as Entries @brandonkelly ?

**Comment by brandonkelly** [comment]
Yeah we would implement it for all element index pages.

**Comment by Mosnar** [comment]
FWIW, we threw together a free plugin that does bulk editing for arbitrary elements with certain fields: https://github.com/venveo/craft-bulkedit/

@brandonkelly Here are some gotchas I ran into while developing this, hopefully it saves you some time!
- Validation gets funky very quickly, especially with the introduction of our "strategies" feature. For example, you can "subtract" the selected elements from an element relation field, which can result in the field no longer validating if there's a minimum requirement. You don't know this is going to be an issue until the element is being edited in the queue.
- I had to scrap custom field and matrix support altogether, as there wasn't a good/clean way of handling when a field tried to reference the element it's on in its template. Most basic custom fields were fine, but I recall having issues with the Typed Link Field plugin. I tried "mocking" an element for the field, but quickly abandoned that route. 
- Non-custom fields (element properties) are actually trickier to handle than custom fields since you can't generate the inputs dynamically (for example, how do we render the author vs the post date?). A template would need to be supplied for each property.
- Field layouts were pretty funky to work with since they're stored in different ways per element. I introduced the concept of **Element Processors** to handle determining where to find field layouts:
https://github.com/venveo/craft-bulkedit/tree/master/src/elements/processors
- There wasn't a great way to determine when a job was completed in the queue in the CP, so I could automatically refresh the element index when the job progressed/finished. This leads clients to thinking the plugin doesn't work or multiple jobs in the queue from multiple attempts
- Paginated element indexes don't maintain previous selections, so even with batch editing, you're bottlenecked by the number of results on the current page. (Related: #4603)
- I really wanted to be able to edit fields across multiple entry types at once. This is do-able, but it makes the editing process a little strange, as the fields are lumped together without any regard for their original layout/tabs and more, you could be editing a field that isn't available on a particular entry type in your batched set of entries.

**Comment by MaximilianSchwarz** [comment]
> Yeah we would implement it for all element index pages.

Any news for this feature?

**Comment by brandonkelly** [comment]
Just changed the milestone to Craft 4, where this will fit nicely with some other element index improvements we’re going to be making.

**Comment by lexislav** [comment]
Would be great if the entry title could be batch edited as well. It would come handy at least for assets (image titles).

**Comment by bencresty** [comment]
Please also make sure it would be possible from within PHP/Controllers to edit/create multiple entries first and save them all together with one sql command. Editing and saving now takes a lot of time when having thousands of records to insert or update!

**Comment by brandonkelly** [comment]
This is still on the list, but worth mentioning that as of Craft 3.7.29, it’s now possible to bulk-set a single value from `resave/*` CLI commands. See #10356 for details.

**Comment by mkrigsman** [comment]
Is editing bulk fields possible Craft ver 4?

**Reply by davorpeic** [reply]
@mkrigsman There is this plugin for bulk edit https://github.com/venveo/craft-bulkedit

**Reply by mkrigsman** [reply]
Thank you for the response. We did look at that plugin but it does not show all the fields. Could we doing something wrong?

**Reply by Mosnar** [reply]
> Thank you for the response. We did look at that plugin but it does not show all the fields. Could we doing something wrong?

Bulk Edit plugin author here - please consider opening an issue on the project repository with details of your problem/question so I can help you without derailing this thread. Thanks!

**Reply by svonlanthen** [reply]
@davorpeic 
- https://github.com/venveo/craft-bulkedit seems to work for Craft3 only
- for Craft4 i found https://plugins.craftcms.com/venveo-bulk-edit?craft4 which seems to work nicely
- nevertheless I hope that Bulk-Edit become Craft-native

@brandonkelly 
- Please proove me wrong if there is a native method that i do not see. 
- I'm aware of https://craftcms.com/knowledge-base/bulk-resaving-elements, but I'm looking for something more client-friendly

**Reply by brandonkelly** [reply]
@svonlanthen It’s on the list for v5!

**Reply by gtwright** [reply]
@brandonkelly Glad to hear this is on the list for v5. Any update on how close we are? 

The bulk edit plugin is meeting our needs nicely on v4, but this is likely to be our blocker for updating to v5.

Thanks!

**Reply by davorpeic** [reply]
@gtwright Hi, not sure if you installed v5 yet, but there is Edit button at the bottom of Entries list and you can bulk edit entries with fields you have visible in the table :)

**Reply by gtwright** [reply]
Thanks, @davorpeic. I do have a dev env of v5 running. From what I can tell, I might call that feature "quick edit" where the fields are exposed as editable in the table, but you still have to edit each one individually. Is that the feature you mean?

What I'm looking for is the ability to apply a filter, select all, then do one edit for all of those entries at once. We somewhat regularly update 100+ entries at once, so even that quick edit feature wouldn't fully meet our needs, since it still involves editing individual records (plus some properties aren't editable through that interface)

**Reply by davorpeic** [reply]
@gtwright Yeah, what you need is a little more advanced and agree that actual feature can be called quick edit. Would be good to have a roadmap regarding that feature if it will be planned or "this is it" from that functionality.

**Comment by foxtrotcz** [comment]
If multiple entries have the same entry type it would make sense to be able to bulk edit multiple entries at once.

**Comment by splendidrob** [comment]
@brandonkelly Just checking back in on this feature. I've just found the "edit" tab in the entries list but this doesn't seem to bulk edit at all, more like a quick edit - plus we cant edit any matrix fields. One of our real problems is bulk updating pages built within a matrix page builder - so for example if we wanted to change the background colour of a banner across every page currently we have to open each page entry and update the colour...it would be an insane time saver if there was a bulk edit option to handle this across all matrix fields!

We tried the Bulk Edit plugin but this seems super buggy and couldn't handle any of our requested edits. 

Any chance this is coming to core as suggested in this thread?

Many thanks :)

**Reply by brandonkelly** [reply]
Right, this is still something we need to get to.

**Reply by arnotenhagen** [reply]
"Our customers have requested this feature multiple times, and the demand is recurring. It would be very useful to have a bulk edit functionality in the CMS to change the values of specific fields across multiple entries. A typical use case would be batch-updating the author for multiple entries."
